rfd/0049-desktop-clipboard.md

```diff
@@ -0,0 +1,197 @@
+---
+authors: Zac Bergquist (zac@goteleport.com)
+state: draft
+---
+
+# RFD 0049 - Desktop Access: Clipboard Support
+
+## What
+
+This RFD defines the high-level goals and architecture for supporing copy and
+paste between a Teleport user's workstation and a remote Windows Desktop.
+
+## Details
+
+### Goals
+
+The primary goal of this RFD is to support basic copy paste of small amounts of
+text between a Teleport user's workstation and a remote desktop. For example,
+copying shell commands, URLs, snippets of log files, etc.
+
+### Non-Goals
+
+The following are all considered out of scope for this RFD:
+
+- large (greater than a few MB) amounts of data
+- file transfer (this is a high priority feature but will implemented separately
+  in order to keep scope reasonable)
+
+### Overview of RDP
+
+#### Introduction
+
+An RDP connection contains a number of "virtual channels" that provide separate
+logical streams of data. Today, we use the _global_ channel for desktop data, and
+the _device redirection_ ("RDPDR") channel to implement a smart card emulator.
+
+Clipboard support for RDP is implmemented as a series of messages sent over a
+dedicated virtual channel called "CLIPRDR". The details of the protocol are specified
+in [MS-RDPECLIP](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpeclip/fb9b7e0b-6db4-41c2-b83c-f889c1ee7688).
+
+The spec often refers to the two clipboards as the _shared clipboard_ and the
+_local clipboard_. The terminology can be confusing, but you can think of the
+_shared clipboard_ as the location that is copied from, and the _local
+clipboard_ as the location that is pasted to. For example, if a Teleport user
+copies text from their workstation and wants to paste it in a remote Windows
+Desktop, then the user is the shared clipboard owner and the desktop is the
+local clipboard owner. The roles are reversed for the other direction. When text
+is copied from the remote desktop and pasted on a user's workstation, the
+Windows Desktop is the shared clipboard owner and the Teleport user is the local
+clipboard owner.
+
+#### Data Types
+
+RDP supports several different types of clipboard data:
+
+1. Generic: opaque data, not manipulated in any way
+2. Palette: a series of RGB-tuples, specially encoded for transport on the wire
+3. Metafile: an application-independent vector format for image transfers
+4. File List: a list of files to be transferred
+5. File Stream: the contents of a file, allowing transfer of specific chunks
+
+Initially, Teleport will only support option 1 - generic clipboard data. Options
```

Comment:
This comment seems to perhaps imply that image copy/pastes will be considered separate from other file-type copy-pastes. Is this the case? Will the file transfer feature not allow image files? Or transfer them as regular binary files, and ignore the Palette/Metafile options?

I think these questions are out of scope for this RFD, but this section may warrant revision given our current understanding of these details.

Reply:
The file list / file stream type will support any file, images included.

I'm not currently aware of any reason why palette or metafile would be used today.